frontend/ui/widget/imageviewer.lua

```diff
@@ -713,42 +735,87 @@ function ImageViewer:onSpread(_, ges)
+    -- (where the image dimensions are many times smaller than the screen),
     -- meaning using the image dimensions here takes less zoom steps to get it back to a sensible size;
     -- *and* large scale factors (where the image dimensions are larger than the screen),
     -- meaning using the screen dimensions here makes zoom steps, again, slightly more potent.
-    local inc
+    -- Note that, in the first case, you're very likely to end up in the first branch instead (i.e., the "snap to" zoom).
     if ges.direction == "vertical" then
-        inc = ges.distance / math.min(Screen:getHeight(), self._image_wg:getCurrentHeight())
+        local img_h = self._image_wg:getCurrentHeight()
+        local screen_h = Screen:getHeight()
+        if (0.9 * ges.span) > img_h then
+            self:onZoomToHeight(ges.span)
+        else
+            self:onZoomIn(ges.distance / math.min(screen_h, img_h))
+        end
     elseif ges.direction == "horizontal" then
-        inc = ges.distance / math.min(Screen:getWidth(), self._image_wg:getCurrentWidth())
+        local img_w = self._image_wg:getCurrentWidth()
+        local screen_w = Screen:getWidth()
+        if (0.9 * ges.span) > img_w then
+            self:onZoomToWidth(ges.span)
+        else
+            self:onZoomIn(ges.distance / math.min(screen_w, img_w))
+        end
     else
+        local img_d = self._image_wg:getCurrentDiagonal()
         local tl = Geom:new{ x = 0, y = 0 }
         local br = Geom:new{ x = Screen:getWidth() - 1, y = Screen:getHeight() - 1}
-        local screen_diag = tl:distance(br)
-        inc = ges.distance / math.min(screen_diag, self._image_wg:getCurrentDiagonal())
+        local screen_d = tl:distance(br)
+        if (0.9 * ges.span) > img_d then
+            self:onZoomToDiagonal(ges.span)
+        else
+            self:onZoomIn(ges.distance / math.min(screen_d, img_d))
+        end
     end
-    self:onZoomIn(inc)
     return true
 end
 
 function ImageViewer:onPinch(_, ges)
     -- With Pinch, unlike Spread, it feels more natural if we keep the same center point.
-    -- Set some zoom decrease value from pinch distance
-    local dec
+    -- As for the actual zoom methods, the same general principle applies,
+    -- except that we don't want to use the "snap to" method when the image is larger than the screen,
+    -- otherwise we'd lose granularity in this case.
+    -- We also prevent "snap to" pinches if it would go below 50% of the current dimension,
+    -- as large "snap to" pinches are much easier to trigger than their matching spread counterpart,
+    -- especially on large screens.
     if ges.direction == "vertical" then
-        dec = ges.distance / math.min(Screen:getHeight(), self._image_wg:getCurrentHeight())
+        local img_h = self._image_wg:getCurrentHeight()
+        local screen_h = Screen:getHeight()
+        if (0.9 * ges.span) < img_h and ges.span > (0.5 * img_h) and img_h < screen_h then
```

Comment:
And these comparisons with various ordering of its items is also hareder to read.
`if ges.span > img_h * 0.5 and ges.span < img_h * 1.1 and img_h < screen_h then` reads easier